producer/writer/writer.go

```diff
@@ -0,0 +1,258 @@
+	}
+	// NB(cw): Need to inc ref here in case a consumer service
+	// writes the data too fast and close the data.
+	d.IncRef()
+	for _, csw := range m {
+		csw.Write(d)
+	}
+	d.DecRef()
+	return nil
+}
+
+func (w *writer) Init() error {
+	w.mPool.Init()
+	newUpdatableFn := func() (watch.Updatable, error) {
+		return w.ts.Watch(w.topic)
+	}
+	getUpdateFn := func(value watch.Updatable) (interface{}, error) {
+		t, err := value.(topic.Watch).Get()
+		if err != nil {
+			w.m.invalidTopicUpdate.Inc(1)
+			return nil, err
+		}
+		return t, nil
+	}
+	vOptions := watch.NewOptions().
+		SetInitWatchTimeout(w.opts.PlacementWatchInitTimeout()).
+		SetInstrumentOptions(w.opts.InstrumentOptions()).
+		SetNewUpdatableFn(newUpdatableFn).
+		SetGetUpdateFn(getUpdateFn).
+		SetProcessFn(w.processFn)
+	w.value = watch.NewValue(vOptions)
+	return w.value.Watch()
+}
+
+func (w *writer) process(update interface{}) error {
+	t := update.(topic.Topic)
+	// We don't allow changing number of shards for topics, it will be
+	// prevented on topic service side, but also being defensive here as well.
+	if w.numShards != 0 && w.numShards != t.NumberOfShards() {
+		w.m.topicUpdateError.Inc(1)
+		return fmt.Errorf("invalid topic update with %d shards, expecting %d", t.NumberOfShards(), w.numShards)
+	}
+	var (
+		iOpts                     = w.opts.InstrumentOptions()
+		newConsumerServiceWriters = make(map[string]consumerServiceWriter, len(t.ConsumerServices()))
+		toBeClosed                []consumerServiceWriter
+		multiErr                  xerrors.MultiError
+	)
+	for _, cs := range t.ConsumerServices() {
+		key := cs.ServiceID().String()
+		csw, ok := w.consumerServiceWriters[key]
+		if ok {
+			newConsumerServiceWriters[key] = csw
+			continue
+		}
+		scope := iOpts.MetricsScope().Tagged(map[string]string{
+			"consumer-service-name": cs.ServiceID().Name(),
+			"consumer-service-zone": cs.ServiceID().Zone(),
+			"consumer-service-env":  cs.ServiceID().Environment(),
+			"consumption-type":      cs.ConsumptionType().String(),
+		})
+		csw, err := newConsumerServiceWriter(cs, t.NumberOfShards(), w.mPool, w.opts.SetInstrumentOptions(iOpts.SetMetricsScope(scope)))
+		if err != nil {
+			w.logger.Errorf("could not create consumer service writer for %s: %v", cs.String(), err)
+			multiErr = multiErr.Add(err)
+			continue
+		}
+		if err = csw.Init(w.initType); err != nil {
+			w.logger.Errorf("could not init consumer service writer for %s: %v", cs.String(), err)
+			multiErr = multiErr.Add(err)
+			csw.Close()
+			continue
+		}
+		newConsumerServiceWriters[key] = csw
+	}
+	for key, csw := range w.consumerServiceWriters {
+		if _, ok := newConsumerServiceWriters[key]; !ok {
+			toBeClosed = append(toBeClosed, csw)
+		}
+	}
+	// Allow InitValueError for any topic updates during runtime.
+	w.initType = allowInitValueError
+	w.m.numConsumerServices.Update(float64(len(newConsumerServiceWriters)))
+
+	// Apply the new consumer service writers.
+	w.Lock()
```

Comment:
Do you still want to apply the update with errors?

Reply:
yeah, if we are adding 2 services during runtime and one has error, we will still add the other success one. This will return err still so that producer.Writer.Init() will fail if there are errors on starting up.